### PR TITLE
Get scipp building and running in VS 2019

### LINKED ISSUE #349: CI for visual c++
This was originally listed here #58 before the big refactor in April.

Though we might not consider deployment on windows, sanitising errors and warnings alone might be worthwhile. Suggested CI [here](https://www.appveyor.com/)

### PR DESCRIPTION
**Pull request by OwenArnold**
Partially fixes #349, though there are further todos to get good PR integration that will be done as a follow up. See [Appveyor Build](https://ci.appveyor.com/project/Scipp/scipp-l7qf3) for branch

**Tester**

Should be able to run all tests, use the python interface etc.

### CODE REVIEW

**Comment by OwenArnold** [comment]
So this seems to be almost there now, all targets built, c++ unit tests complete properly, and I have the builds running on `appveyor` to show this see [fork](https://ci.appveyor.com/project/OwenArnold/scipp) based on these changes.

The remaining problem seems to be a problem in scipp. You can see python errors [here](https://ci.appveyor.com/project/OwenArnold/scipp#L6292) 
```
    def test_create():
>       x = sc.Variable([Dim.X], values=np.arange(3))
E       RuntimeError: Unsupported numpy dtype.
tests\test_dataset.py:18: RuntimeError

```
From here https://github.com/scipp/scipp/blob/master/python/dtype.cpp#L32

On my macbook:
```
>>> type(np.arange(1)[0])
<class 'numpy.int64'>
```
On my windows 10 machine:
```
>>> type(np.arange(1)[0])
<class 'numpy.int32'>
```
I think this is the same issue discussed [here](https://github.com/pybind/pybind11/issues/563), where they promote using `py::isinstance<py::array_t<T>>(o)` rather than `dtype::of` as is currently done. So there is perhaps a genuine issue in scipp with this that needs to be fixed? Thoughts please @SimonHeybrock

**Comment by OwenArnold** [comment]
Update: Looking at https://github.com/scipp/scipp/pull/587, this issue might have been fixed? I'll ensure I have these changes.

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,20 +26,21 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/CMake"
...
 
+include(GTest)
 include(boost)
 include(Eigen)
 include(pybind11)
-include(GoogleTest)
 include(GoogleBenchmark)
 include(GenerateExportHeader)
+find_package(Boost REQUIRED)
+find_package(Python COMPONENTS Interpreter Development)
```
Comment (DTasev) [inline CMakeLists.txt]:
```suggestion
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -27,20 +26,21 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/CMake"
...
 
-if("${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}" VERSION_LESS "3.5")
-  message(FATAL_ERROR "Python v3 interpreter must be greater than or equal to 3.5. Found ${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}")
+if("${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}" VERSION_LESS "3.5")
```
Comment (DTasev) [inline CMakeLists.txt]:
```suggestion
if("${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}" VERSION_LESS "3.5")
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -27,20 +26,21 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/CMake"
...
+  message(FATAL_ERROR "Python v3 interpreter must be greater than or equal to 3.5. Found ${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}")
```
Comment (DTasev) [inline CMakeLists.txt]:
```suggestion
  message(FATAL_ERROR "Python v3 interpreter must be greater than or equal to 3.5. Found ${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR}")
```

File: CMake/boost.cmake [hunk 4]
```diff
@@ -25,4 +25,6 @@ if(result)
   message(FATAL_ERROR "Install step for boost failed: ${result}")
 endif()
 
-set(BOOST_ROOT "${CMAKE_BINARY_DIR}/boost/include" CACHE PATH "Path to boost")
+set(BOOST_INCLUDEDIR "${CMAKE_BINARY_DIR}/boost/include" CACHE PATH "Path to boost")
+set(BOOST_NO_SYSTEM_PATHS ON)
+#set(Boost_INCLUDE_DIRS "${CMAKE_BINARY_DIR}/boost/include")
```
Comment (DTasev) [inline CMake/boost.cmake]:
```suggestion
```

File: core/include/scipp/core/variable.tcc [hunk 5]
```diff
@@ -10,9 +10,14 @@
 #include "scipp/core/variable_view.h"
 #include "scipp/core/vector.h"
 #include "scipp/units/unit.h"
+#include <optional>
```
Comment (DTasev) [inline core/include/scipp/core/variable.tcc]: Unused header?

File: CMakeLists.txt [hunk 6]
```diff
@@ -27,20 +26,21 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/CMake"
...
 endif()
```
Comment (DTasev) [inline CMakeLists.txt]: Some empty lines left below here

File: appveyor.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+- git submodule update --init
+- mkdir build
+- cd build%
+
+build_script:
+- cmake -G "Visual Studio 16 2019" -A x64 -DPYTHON_EXECUTABLE="C:/Python36-x64/python.exe" -DCMAKE_INSTALL_PREFIX=../install ..
+- cmake --build . --config %CONFIGURATION% --target install all-tests
+
+test_script:
+- ctest -VV
```
Comment (SimonHeybrock) [inline appveyor.yml]: I think we should probably not be using `ctest` to run tests. As far as I remember it had trouble picking up some templated tests, i.e., not all tests were being run.

File: python/variable.cpp [hunk 8]
```diff
@@ -193,7 +193,12 @@ void init_variable(py::module &m) {
   // This should be in the certain order
   bind_init_0D(variable);
   bind_init_0D<bool>(variable);
+#ifdef _WIN32
+  // C long 32 bit on windows.
+  bind_init_0D<int32_t>(variable);
```
Comment (SimonHeybrock) [inline python/variable.cpp]: Are Python integers 32 bit on Windows?

File: python/tests/test_variable.py [hunk 9]
```diff
@@ -589,7 +590,11 @@ def test_construct_0d_numpy():
 
 
 def test_construct_0d_native_python_types():
-    assert sc.Variable(2).dtype == sc.dtype.int64
+    if sys.platform == "win32":
+        # Windows defaults to a 32 bit integer
+        assert sc.Variable(2).dtype == sc.dtype.int32
+    else:
+        assert sc.Variable(2).dtype == sc.dtype.int64
```
Comment (SimonHeybrock) [inline python/tests/test_variable.py]: I don't understand, this does not make sense. Why?

File: python/dtype.cpp [hunk 10]
```diff
@@ -25,7 +24,8 @@ scipp::core::DType scipp_dtype(const py::dtype &type) {
   if (type.is(py::dtype::of<std::int64_t>()) ||
       (type.kind() == 'i' && type.itemsize() == 8))
     return scipp::core::dtype<int64_t>;
-  if (type.is(py::dtype::of<std::int32_t>()))
+  if (type.is(py::dtype::of<std::int32_t>()) ||
+      (type.kind() == 'i' && type.itemsize() == 4))
```
Comment (SimonHeybrock) [inline python/dtype.cpp]: Why is this needed now?

File: appveyor.yml (hunk 7 above)
Comment (OwenArnold) [reply appveyor.yml]: Unfortunately I think you are right different total test count between appveyor and travis. I'll change this.

File: python/variable.cpp (hunk 8 above)
Comment (OwenArnold) [reply python/variable.cpp]: Yes https://docs.microsoft.com/en-us/cpp/c-language/storage-of-basic-types?redirectedfrom=MSDN&view=vs-2019 https://stackoverflow.com/questions/22513445/python-handles-long-ints-differently-on-windows-and-unix

File: python/tests/test_variable.py (hunk 9 above)
Comment (OwenArnold) [reply python/tests/test_variable.py]: As discussed above.  `long` 4 bytes windows, 8 bytes linux.

Comment (SimonHeybrock) [reply python/tests/test_variable.py]: 😱

Comment (SimonHeybrock) [reply python/tests/test_variable.py]: (I mean, I knew that this is what C/C++ do, but I had assumed Python would not follow this track)

File: python/dtype.cpp (hunk 10 above)
Comment (OwenArnold) [reply python/dtype.cpp]: All linked to the above. i.e on windows doing something like:
```python
var = Variable([Dim.X], np.arange(10))
```
exposes this problem.

File: core/include/scipp/core/variable.tcc (hunk 5 above)
Comment (OwenArnold) [reply core/include/scipp/core/variable.tcc]: No this should be included.

Comment (DTasev) [reply core/include/scipp/core/variable.tcc]: I am not sure what is used from it, is it `std::bad_variant_access`? If so it might be clearer to include `<variant>` instead?

Comment (OwenArnold) [reply core/include/scipp/core/variable.tcc]: https://github.com/scipp/scipp/blob/6b5dd66e50e1ef3fef417be08106496c8f8e9bb3/core/include/scipp/core/variable.tcc#L53

File: python/variable.cpp (hunk 8 above)
Comment (DTasev) [reply python/variable.cpp]: Are we sure that the `int64` constructor is not needed on Windows? This makes it impossible to store a 0D value larger than `2<<30 - 1`, so this `scipp.Variable(2<<30)` fails as it cannot find a representation. I doubt many things go that high, but maybe some log values could (e.g. timestamps).

Having both int32 and int64 seems to work OK:

```
In [2]: scipp.Variable(1)
Out[2]:     <scipp.Variable>          int32      [dimensionless]  ()  [1]

In [3]: scipp.Variable(2<<30)
Out[3]:     <scipp.Variable>          int64      [dimensionless]  ()  [2147483648]
```

It is however dependent on the order it is defined in the source. If the bind_init_0D<int64_t> is defined first then `scipp.Variable(1)` will bind to `int64`..

Comment (OwenArnold) [reply python/variable.cpp]: @DTasev instead of the compile-time or overloaded route - maybe there is a smarter way to do this at runtime based on the size of the integer passed? I believe this is how python and numpy interoperate.

Comment (SimonHeybrock) [reply python/variable.cpp]: I would object to runtime based size checks. Can't we just make it `int64_t` and do away with some of the special case handling? I would value *consistency within scipp* **higher** than *consistency with platform-specific Python/numpy behavior*.

Comment (SimonHeybrock) [reply python/variable.cpp]: ... in particular since no significant harm can come from using a *wider* type.

File: core/include/scipp/core/variable.tcc (hunk 5 above)
Comment (DTasev) [reply core/include/scipp/core/variable.tcc]: Ah okay, I thought it was related to the new additions. That makes snese

File: python/variable.cpp (hunk 8 above)
Comment (DTasev) [reply python/variable.cpp]: `int64_t` can be made the default, the issue is - then every `np.arange` needs to have `dtype=np.int64` added, or tests that do `sc.Variable(N)` will fail their equality checks on Windows. 

(By the way `np.ones` gives `float64` by default on Windows, `numpy` is not fully consistent with itself either!)

File: CMakeLists.txt (hunk 1 above)
Comment (SimonHeybrock) [reply CMakeLists.txt]: @OwenArnold Why is this needed?

File: CMakeLists.txt [hunk 11]
```diff
@@ -27,21 +26,19 @@ set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/CMake"
...
 find_package(Sanitizers REQUIRED)
 
+include(GTest)
 include(boost)
 include(Eigen)
 include(pybind11)
-include(GoogleTest)
 include(GoogleBenchmark)
 include(GenerateExportHeader)
+find_package(Boost REQUIRED)
```
Comment (SimonHeybrock) [inline CMakeLists.txt]: Any reasons for dropping the `1.67` requirement? We do actually need that.

Comment (OwenArnold) [reply CMakeLists.txt]: Yes, that can be added back.

**Comment by SimonHeybrock** [comment]
Thanks a lot @OwenArnold for all the work that went into this!
